lib/procbot.ts

```diff
@@ -0,0 +1,265 @@
+
+/**
+ * A map linking contexts to Worker instances.
+ * @typedef {Object} ProcBot.WorkerMap
+ * @property {generic} - A generic context for the Worker.
+ * @property {Worker}
+ */
+type WorkerMap<T> = Map<T, Worker<T>>;
+
+/**
+ * The Worker class is responsible for the execution of scheduling tasks based on events.
+ * @class ProcBot.Worker
+ * @classdesc
+ * Each Worker instance is bound to a context. This context could be, for example, a
+ * unique Github repository, or a directory in a file system, a specific customer service
+ * in a set, etc. It is also generic, and can therefore be of any type.
+ * When WorkerEvents are added to an empty queue, they are processed for being worked on
+ * in the next tick of event loop.
+ */
+export class Worker<T> {
+    /**
+     * Holds the context for the Worker.
+     * @member Worker._context
+     * @private
+     * @type {generic}
+     */
+    private _context: T;
+    /**
+     * Reference to the parent Map in which this Worker exists.
+     * @member Worker.parentMap
+     * @private
+     * @type {generic}
+     */
+    private parentMap: WorkerMap<T>;
+    /**
+     * Holds the queue of events to work on.
+     * @member Worker.queue
+     * @private
+     * @type {ProcBot.WorkerEvent[]}
+     */
+    private queue: WorkerEvent[] = [];
+
+    /**
+     * Creates the Worker class, specifying a context and the parent Map.
+     * @param {generic} context - The context to use for hashing.
+     * @param {ProcBot.WorkerMap} parentMap - The parent Map containing all Workers.
+     */
+    constructor(context: T, parentMap: WorkerMap<T>) {
+        this._context = context;
+        this.parentMap = parentMap;
+    }
+
+    /**
+     * Retrieve the context for the Worker.
+     * @return {generic} - The context.
+     */
+    get context() {
+        return this._context;
+    }
+
+    /**
+     * Add a new event to the Worker's event queue.
+     * @param {ProcBot.WorkerEvent} worker - The event to add to the queue.
+     */
+    addEvent(event: WorkerEvent) {
+        this.queue.push(event);
+        // If this is a new worker, ensure it operates.
+        if (this.queue.length === 1) {
+            this.runWorker();
+        }
+    }
+
+    // Run as many workers as are queued. Do this atomically, in FIFO order.
+    /**
+     * Runs the next worker in the queue, before deleting its entry. Should more entries
+     * exist it then runs those.
+     */
+    private runWorker() {
+        // Get the next thing from the queue.
+        const entry = <WorkerEvent>this.queue.shift();
+        const self: this = this;
+
+        // Run worker, proceed to next worker.
+        entry.workerMethod(entry.event, entry.data)
+        .then(() => {
+            if (this.queue.length > 0) {
+                process.nextTick(this.runWorker);
+            } else {
+                // Unlink ourselves from our parent list.
+                self.parentMap.delete(self.context);
+            }
+        });
+    }
+}
```

Comment:
You're right, TypeScript doesn't really have a concept of module privacy - everything is file scoped, and you can't have something (afaik) that's shared between multiple files but otherwise private without doing some very hacky things with inheritance, which I'd avoid.

I think the prevailing logic is that this is fine though. If you have two entities that are meaningful standalone, and they share a common model, then that model is reasonably meaningful standalone in your codebase too, and should be public. Since Worker is an exported class and this is the type of its constructor argument, WorkerMap is more or less part of your domain anyway. Alternatively, you're sharing an implementation detail here that other classes shouldn't even know about, and it actually shouldn't be shared at all.

Some interesting solutions:

* Don't tell Worker about the WorkerMap. It currently only uses it to remove itself from it, but arguably that's it having too much knowledge about how it's used. You could just pass an `onDone` callback in when it's created, which calls some `.removeWorker` method to unregister itself. Then all it knows about is the callback it's given, the procbot subclasses just know about the add/get/remove worker methods, and ProcBot itself is the only one who knows there's a map involved. **I think this is actually probably the right option here.**
* Just make more things public and don't worry about it.
* Make things public, but use the structure to make things clearer. You could have lib/procbot/procbot.ts, lib/procbot/worker.ts, lib/procbot/internals.ts, which clearly points people to the right bits of the code, but doesn't require you to bother enforcing it.
* Duplicate type between the Worker and ProcBot modules. Feels nasty, but there's no concrete downside I think, since structural typing means they'll still be considered the same, and type checking will stop you ever going wrong. Also lets them evolve independently - you can have extra information in the map that ProcBot uses, without Workers finding out about it.